### PR TITLE
Implement Custom Labels for PRs

### PR DESCRIPTION
**Pull request by hussam789**
## PR Type:
Enhancement

___
## PR Description:
This PR introduces the ability to use custom labels for PRs. The main changes include:
- Addition of a new command `/generate_labels` for generating PR labels.
- Custom labels can be defined in the `custom_labels.toml` settings file.
- Custom labels are disabled by default and can be enabled in the `configuration.toml` settings file.
- The PR description and review prompts have been updated to support custom labels.
- A new class `PRGenerateLabels` has been added to handle the generation of PR labels.
- Environment variables `CUSTOM_LABELS` and `CUSTOM_LABELS_DESCRIPTION` have been added to the GitHub action runner to handle custom labels.

___
## PR Main Files Walkthrough:
<details> <summary>files:</summary>

`pr_agent/agent/pr_agent.py`: Added the `PRGenerateLabels` class to the list of commands.
`pr_agent/algo/utils.py`: Added a `set_custom_labels` function to handle the setting of custom labels.
`pr_agent/config_loader.py`: Added `pr_custom_labels.toml` to the list of settings files to be loaded.
`pr_agent/servers/github_action_runner.py`: Added handling for the `CUSTOM_LABELS` and `CUSTOM_LABELS_DESCRIPTION` environment variables.
`pr_agent/tools/pr_description.py`: Updated the variables dictionary to include custom labels related variables.
`pr_agent/tools/pr_generate_labels.py`: Added a new class `PRGenerateLabels` to handle the generation of PR labels.
`pr_agent/tools/pr_reviewer.py`: Updated the variables dictionary to include the custom labels variable.
`pr_agent/settings/configuration.toml`: Added a new setting `enable_custom_labels` to enable or disable custom labels.
`pr_agent/settings/custom_labels.toml`: Added a new settings file for defining custom labels.
`pr_agent/settings/pr_custom_labels.toml`: Added a new settings file for defining the prompts for generating custom labels.
</details>

### CODE REVIEW

File: pr_agent/tools/pr_reviewer.py [hunk 1]
```diff
@@ -149,6 +150,7 @@ async def _get_prediction(self, model: str) -> str:
         variables["diff"] = self.patches_diff  # update diff
 
         environment = Environment(undefined=StrictUndefined)
+        await set_custom_labels(variables)
```
Comment (mrT23) [inline pr_agent/tools/pr_reviewer.py:153]: why is this async ?

File: pr_agent/settings/configuration.toml [hunk 2]
```diff
@@ -33,10 +33,30 @@ add_original_user_description=false
 keep_original_user_title=false
 use_bullet_points=true
 extra_instructions = ""
+
 # markers
 use_description_markers=false
 include_generated_by_header=true
 
+[custom_labels."Bug fix"]
```
Comment (mrT23) [inline pr_agent/settings/configuration.toml:41]: this takes a lot of room in the main configuration file

should move to a separate file

File: pr_agent/settings/configuration.toml [hunk 3]
```diff
@@ -33,10 +33,30 @@ add_original_user_description=false
...
+[custom_labels."Bug fix with tests"]
+description = "Fixes a bug in the code and adds or modifies tests"
+[custom_labels."Refactoring"]
+description = "Refactors the code without changing its functionality"
+[custom_labels."Enhancement"]
+description = "Adds new features or functionality"
+[custom_labels."Documentation"]
+description = "Adds or modifies documentation"
+[custom_labels."SQL modifications"]
+description = "Adds or modifies SQL queries"
```
Comment (mrT23) [inline pr_agent/settings/configuration.toml:54]: this seems like overfit to SQL

File: pr_agent/settings/pr_description_prompts.toml [hunk 4]
```diff
@@ -19,16 +19,11 @@ PR Title:
   description: an informative title for the PR, describing its main theme
 PR Type:
   type: array
+  description: One or more labels that describe the PR type. Don't output the description in the parentheses.
```
Comment (mrT23) [inline pr_agent/settings/pr_description_prompts.toml:22]: this should be optional, only when 'custom_labels mode is enabled

File: pr_agent/servers/github_action_runner.py [hunk 5]
```diff
@@ -33,13 +36,36 @@ async def run_action():
     if not GITHUB_TOKEN:
         print("GITHUB_TOKEN not set")
         return
+    if CUSTOM_LABELS:
```
Comment (mrT23) [inline pr_agent/servers/github_action_runner.py:39]: maybe move all this to a seperate dedicated function ?

File: pr_agent/algo/utils.py [hunk 6]
```diff
@@ -304,3 +304,18 @@ def try_fix_yaml(review_text: str) -> dict:
         except:
             pass
     return data
+
+
+async def set_custom_labels(variables):
```
Comment (mrT23) [inline pr_agent/algo/utils.py:309]: why async ?

Comment (hussam789) [reply pr_agent/algo/utils.py:309]: Because it is called from an async function

File: pr_agent/settings/configuration.toml (hunk 3 above)
Comment (hussam789) [reply pr_agent/settings/configuration.toml:54]: I will remove the SQL version ( it should not be part of the default labels)

File: pr_agent/settings/configuration.toml (hunk 2 above)
Comment (mrT23) [reply pr_agent/settings/configuration.toml:41]: No
move that big chunk of text to another file.
its unreasonable to dedicated so many lines in a general configuration to a tailor made optional feature.

**Comment by hussam789** [comment]
Preparing PR labels...

**Comment by hussam789** [comment]
Preparing PR description...

**Comment by hussam789** [comment]
Preparing PR labels...

**Comment by hussam789** [comment]
Preparing PR labels...

**Review by mrT23 — APPROVED** [review]

**Comment by hussam789** [comment]
Preparing review...

**Comment by hussam789** [comment]
## PR Analysis

- 🎯 **Main theme:** Implementing custom labels for PRs
- 📝 **PR summary:** This PR introduces the ability to use custom labels for PRs. The main changes include the addition of a new command `/generate_labels` for generating PR labels, the ability to define custom labels in the `custom_labels.toml` settings file, and the addition of environment variables `CUSTOM_LABELS` and `CUSTOM_LABELS_DESCRIPTION` to handle custom labels.
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** No
- ⏱️ **Estimated effort to review [1-5]:** 4, due to the large number of changes across multiple files and the introduction of new functionalities.
- 🔒 **Security concerns:** No security concerns found
## PR Feedback

- 💡 **General suggestions:** The PR is well-structured and the changes are logically grouped. However, it would be beneficial to include tests for the new functionalities introduced, especially for the `/generate_labels` command and the handling of custom labels.


- **<details><summary> 🤖 Code feedback:**</summary>


  - **relevant file:** ``pr_agent/servers/github_action_runner.py``
   **suggestion:** Consider adding error handling for the case where the `CUSTOM_LABELS` and `CUSTOM_LABELS_DESCRIPTION` environment variables do not have the same length. This could lead to a mismatch between labels and their descriptions. [important]
   **relevant line:** [CUSTOM_LABELS_DICT = handle_custom_labels(CUSTOM_LABELS, CUSTOM_LABELS_DESCRIPTION)](https://github.com/Codium-ai/pr-agent/pull/396/files#diff-15f9c11da031449998128bdbba057d768853e836eedf44563be71b3847be80e0R39)


  - **relevant file:** ``pr_agent/algo/utils.py``
   **suggestion:** It would be beneficial to add a check to ensure that the `labels` variable is not empty before proceeding to use it. This would prevent potential errors if the `labels` variable is not properly initialized. [medium]
   **relevant line:** [labels = get_settings().custom_labels](https://github.com/Codium-ai/pr-agent/pull/396/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53R310)


  - **relevant file:** ``pr_agent/tools/pr_generate_labels.py``
   **suggestion:** Consider adding a check to ensure that the `pr_types` variable is not empty before returning it. This would prevent potential errors if the `pr_types` variable is not properly initialized. [medium]
   **relevant line:** [return pr_types](https://github.com/Codium-ai/pr-agent/pull/396/files#diff-15739956ea92e33af13af3d1725ff1416bca4b087dadacb8aa1c1a9dd993f1c7R163)


  - **relevant file:** ``pr_agent/settings/configuration.toml``
   **suggestion:** It would be beneficial to add a comment explaining what the `enable_custom_labels` setting does. This would improve the readability and maintainability of the code. [medium]
   **relevant line:** [enable_custom_labels=false](https://github.com/Codium-ai/pr-agent/pull/396/files#diff-66cfda5143e484ee53ecf7aa0df7dca8ad0b181256f4b0675905db35bcbbae78R40)

</details>

**Comment by hussam789** [comment]
Preparing PR labels...

**Comment by hussam789** [comment]
Preparing PR labels...

**Pull request merged**